feat: fetcher from clipboard and auto-select identifier type if valid

lydia-yan opened the pull request:
feat: fetcher from clipboard and auto-select identifier type if valid
Update https://github.com/JabRef/jabref/pull/13091 https://github.com/JabRef/jabref/issues/13087

### Summary of changes
This PR improves the user experience when creating a new entry by automatically detecting the identifier type (DOI, ISBN, arXiv, RFC, SSRN) from the clipboard content and selecting the appropriate fetcher in the "Enter Identifier" tab.

Changes made:
- Clipboard content is parsed via `CompositeIdFetcher.getIdentifier(...)`
- Added logic to reject suspicious identifiers (e.g., raw numbers like `123456789`)
- If a valid identifier is detected, the dialog switches to "Specify identifier type"
- The corresponding fetcher is auto-selected (e.g., DoiFetcher, IsbnFetcher, etc.)

Unit tests verify identifier detection and validation logic.
UI switching (radio selection and fetcher assignment) was manually verified using valid/invalid identifier clipboard content. 

### Screenshots
Before (macOS):
Clipboard with `10.1109/MCOM.2010.5673082` → No tab switch, user must manually select type  
![image](https://github.com/user-attachments/assets/2420e39f-07f6-41a9-be0a-b9c5212b7a7f)

After (macOS): 
Clipboard with `10.1109/MCOM.2010.5673082` → Automatically switches and pre-selects to DOI
<img width="988" alt="image" src="https://github.com/user-attachments/assets/7c81fdbb-09ae-48c3-90de-e946a9b0fdc6" />

### Mandatory checks
- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (if change is visible to the user)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): No update needed
- [x] [Checked documentation](https://docs.jabref.org/): No update needed

### Collaborators
[@lydia-yan](https://github.com/lydia-yan) [@yoasaaa](https://github.com/yoasaaa) [@brandon-lau0](https://github.com/brandon-lau0) [@FlyJoanne](https://github.com/FlyJoanne)

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. We use [OpenRewrite](https://docs.openrewrite.org/) to ensure "modern" Java coding practices. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / OpenRewrite (pull_request)" and click on it.

The issues found can be **automatically fixed**. Please execute the gradle task *`rewriteRun`* from the [`rewrite` group of the Gradle Tool window](https://devdocs.jabref.org/code-howtos/faq.html#failing-openrewrite-tests) in IntelliJ, then check the results, commit, and push.


<!-- ghprcomment
Tests
OpenRewrite
-->

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. We use [OpenRewrite](https://docs.openrewrite.org/) to ensure "modern" Java coding practices. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / OpenRewrite (pull_request)" and click on it.

The issues found can be **automatically fixed**. Please execute the gradle task *`rewriteRun`* from the [`rewrite` group of the Gradle Tool window](https://devdocs.jabref.org/code-howtos/faq.html#failing-openrewrite-tests) in IntelliJ, then check the results, commit, and push.


<!-- ghprcomment
Tests
OpenRewrite
-->

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. We use [OpenRewrite](https://docs.openrewrite.org/) to ensure "modern" Java coding practices. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / OpenRewrite (pull_request)" and click on it.

The issues found can be **automatically fixed**. Please execute the gradle task *`rewriteRun`* from the [`rewrite` group of the Gradle Tool window](https://devdocs.jabref.org/code-howtos/faq.html#failing-openrewrite-tests) in IntelliJ, then check the results, commit, and push.


<!-- ghprcomment
Tests
OpenRewrite
-->

koppor commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -273,6 +270,45 @@ private void initializeLookupIdentifier() {
         }
 
+        if (!StringUtil.isBlank(clipboardText) && !clipboardText.contains("\n")) {
+            // :TODO: Better validation would be nice here, so clipboard text is only copied over if it matches a
+            // supported identifier format.
```
Yes - just move it to the `isvalid` section?`

General thing: The wish is that when JabRef is ready to expect user input, the currently visible text field should contain the clipboard contents and be selected. - Don't know there in the code this is exactly implemented. Currently, it seems to implemented for each tab; but it could also be implemetned after the whole initalization is run - the input text field of the selected tab should be focused.

This could be a good excercise to train Requirements Tracking - see https://devdocs.jabref.org/requirements/ - It could be placed below Focus (https://devdocs.jabref.org/requirements/focus.html) or as a seperate markdown document for the new-entry-dialog.

koppor commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -273,6 +270,45 @@ private void initializeLookupIdentifier() {
+                        idLookupSpecify.setSelected(true);
+                        for (IdBasedFetcher fetcher : idFetcher.getItems()) {
+                            if ((id instanceof DOI && fetcher instanceof DoiFetcher) ||
+                                    (id instanceof ISBN && fetcher.getName().toLowerCase().contains("isbn")) ||
```
second part `fetcher instanceof AbstractIsbnFetcher` (or comment why you do a string matching)

koppor commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -273,6 +270,45 @@ private void initializeLookupIdentifier() {
+                        for (IdBasedFetcher fetcher : idFetcher.getItems()) {
+                            if ((id instanceof DOI && fetcher instanceof DoiFetcher) ||
+                                    (id instanceof ISBN && fetcher.getName().toLowerCase().contains("isbn")) ||
+                                    (id instanceof ArXivIdentifier && fetcher.getName().toLowerCase().contains("arxiv")) ||
```
second part - `instanceof `ArXiVFetcher || instanceof ArXiV` (if the latter is needed, IDK)

koppor commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -273,6 +270,45 @@ private void initializeLookupIdentifier() {
+                            if ((id instanceof DOI && fetcher instanceof DoiFetcher) ||
+                                    (id instanceof ISBN && fetcher.getName().toLowerCase().contains("isbn")) ||
+                                    (id instanceof ArXivIdentifier && fetcher.getName().toLowerCase().contains("arxiv")) ||
+                                    (id instanceof RFC && fetcher.getName().toLowerCase().contains("rfc")) ||
```
Why no instanceof for `RfcFetcher`?

koppor commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -273,6 +270,45 @@ private void initializeLookupIdentifier() {
+                if (isValid) {
+                    Platform.runLater(() -> {
+                        idLookupSpecify.setSelected(true);
+                        for (IdBasedFetcher fetcher : idFetcher.getItems()) {
```
Move that to a method `fetcherForIdentifier` -- see at the end of the class `fetcherFromName` as another example

koppor commented on jabgui/src/test/java/org/jabref/gui/newentry/NewEntryViewIdentifierDetectionTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test
+    void detectsValidDOI() {
+        Optional<Identifier> result = CompositeIdFetcher.getIdentifier("10.1109/MCOM.2010.5673082");
```
Better include the tests in `org.jabref.logic.importer.fetcher.CompositeIdFetcherTest` to have it more closerer to the method.

(Maybe, the whole `getIdentifier`  should move to org.jabref.model.entry.identifier.Identifier as `from(String identifier))

jabref-machine commented:
Your pull request conflicts with the target branch.

Please [merge `upstream/main`](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-branch-from-the-command-line) with your code. For a step-by-step guide to resolve merge conflicts, see <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line>.


<!-- ghprcomment
On PR opened/updated
Conflicts with target branch
-->

lydia-yan commented:
Hi @koppor,   I've implemented the identifier detection and fetcher matching logic in NewEntryView, and relocated the relevant unit tests to CompositeIdFetcherTest as discussed.

Regarding your suggestion to move getIdentifier() to Identifier.from(String), I wanted to clarify: do you envision this as a static factory method within the Identifier class? If so, would you prefer this method to return an Optional<Identifier>—to indicate possible failure—or should it return a concrete subclass such as DOI or ISBN based on the input? I'd appreciate your guidance on this before proceeding. Thank you！

jabref-machine commented:
You modified Markdown (`*.md`) files and did not meet JabRef's rules for consistently formatted Markdown files. To ensure consistent styling, we have [markdown-lint](https://github.com/DavidAnson/markdownlint) in place. [Markdown lint's rules](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#rules) help to keep our Markdown files consistent within this repository and consistent with the Markdown files outside here.

You can check the detailed error output by navigating to your pull request, selecting the tab "Checks", section "Tests" (on the left), subsection "Markdown".


<!-- ghprcomment
Tests
Markdown
-->

koppor commented:
> Hi @koppor, I've implemented the identifier detection and fetcher matching logic in NewEntryView, and relocated the relevant unit tests to CompositeIdFetcherTest as discussed.

nice!

> Regarding your suggestion to move getIdentifier() to Identifier.from(String)

I need to find the comment again. Maybe, next time, please just put a link to the comment. GitHub is pretty good in deep links.. Reason: I am commenting on dozens of issues and pull requests and don't remember all details. - I did not find the comment, thus, I have to guess.

> , I wanted to clarify: do you envision this as a static factory method within the Identifier class? If so, would you prefer this method to return an Optional—to indicate possible failure—or should it return a concrete subclass such as DOI or ISBN based on the input? I'd appreciate your guidance on this before proceeding. Thank you！

Just move the method from org.jabref.logic.importer.CompositeIdFetcher to org.jabref.model.entry.identifier.Identifier and rename it from `getIdentifier` to `from`. - The class already returns a concrete subclass. The java inheritance ensrues that a concrete class is returned....

koppor commented:
@lydia-yan Thank you for adressing OpenFastTrace! Looks good.

Siedlerchr commented on jabgui/src/main/java/org/jabref/gui/newentry/NewEntryView.java:
```diff
@@ -523,4 +540,43 @@ private static PlainCitationParserChoice parserFromName(String parserName, List<
+        for (IdBasedFetcher fetcher : idFetcher.getItems()) {
+            if ((id instanceof DOI && fetcher instanceof DoiFetcher) ||
+                    // Use instanceof for structured check; fallback to name matching since some ISBN fetchers (like IsbnFetcher) don't extend AbstractIsbnFetcher.
+                    (id instanceof ISBN && (fetcher instanceof AbstractIsbnFetcher ||
```
It is enough if you just do an instanceof ISBNFetcher, because it calls the AbstractISBNFetcher under the hood and will try them one after another if no entry is found.

https://github.com/JabRef/jabref/blob/3b892c0ec1ce569374660e31de08efb36e958cf2/jablib/src/main/java/org/jabref/logic/importer/fetcher/isbntobibtex/IsbnFetcher.java#L66-L83

lydia-yan replied:
Thanks! I've updated the condition to use `instanceof IsbnFetcher` as suggested, and exported the package in `module-info.java` to resolve visibility issues.

lydia-yan commented:
> Just move the method from org.jabref.logic.importer.CompositeIdFetcher to org.jabref.model.entry.identifier.Identifier and rename it from `getIdentifier` to `from`. - The class already returns a concrete subclass. The java inheritance ensrues that a concrete class is returned....

@koppor Thanks! I've moved the method to `Identifier.from(String)` and updated all call sites accordingly. Let me know if you'd like any further refinements.

koppor commented on jablib/src/test/java/org/jabref/logic/importer/fetcher/CompositeIdFetcherTest.java:
```diff
@@ -127,4 +135,41 @@ void performSearchByIdReturnsEmptyForInvalidId(String groundInvalidArXivId) thro
+    @Test
+    void detectsValidISBN() {
+        Optional<Identifier> result = Identifier.from("9780134685991");
+        assertTrue(result.isPresent());
+        assertInstanceOf(ISBN.class, result.get());
+        assertTrue(((ISBN) result.get()).isValid());
```
I think, in this case, it is OK to "unzip" the `Optional` checkings.
